### ISSUE TITLE
Avoid skipping all instance traits with "_items" suffix in HasTraits.traits

### ISSUE THREAD

**Issue by kitchoi** [enthought/traits#1329]
When a trait `List`, `Set` and `Dict` are added to an HasTraits object, and `name_items` trait is added along with it for supporting listening to mutations via `on_trait_change`.

Then in `HasTraits.traits`, instance traits with named ending with "_items" are skipped, persumably because it would otherwise causes `visible_traits` and `editable_traits` to report those `name_items` traits that are added for `on_trait_change` and for use cases like constructing the default TraitsUI view, would be undesirable.

However, this means anyone trying to define a trait with a suffix "_items" is going to be surprised by not being able to get it back via `HasTraits.traits`.

To fix this, I'd propose:
(1) Add `editable` and `visible` metadata `False` for the trait created for supporting listening to `List`, `Set`, and `Dict` mutation.
I believe that is to set `editable=True` in the trait definition in these three places:
https://github.com/enthought/traits/blob/95a8c698ed43a4f44fc336ed62b475c5564ee886/traits/trait_types.py#L2499-L2506
https://github.com/enthought/traits/blob/95a8c698ed43a4f44fc336ed62b475c5564ee886/traits/trait_types.py#L2711-L2717
https://github.com/enthought/traits/blob/95a8c698ed43a4f44fc336ed62b475c5564ee886/traits/trait_types.py#L2847-L2852

(2) Remove this filter in `HasTraits.traits`:
https://github.com/enthought/traits/blob/95a8c698ed43a4f44fc336ed62b475c5564ee886/traits/has_traits.py#L2949

**Comment by kitchoi** [comment]
Actually, this is related to https://github.com/enthought/traits/issues/1098 to some extent.

Consider this:
```
>>> from traits.api import *
>>> 
>>> class Foo(HasTraits):
...     pass
... 
>>> foo = Foo()
>>> foo.add_trait("names", List())
>>> foo.traits()
{'trait_added': <traits.ctrait.CTrait object at 0x7fc96c4deae0>, 'trait_modified': <traits.ctrait.CTrait object at 0x7fc96c4deb80>, 'names': <traits.ctrait.CTrait object at 0x7fc96c5802c0>}
```
The `names_items` trait associated with `List()` is defined on the object, but `HasTraits.traits` has skipped it.

One may take the returned value of `HasTraits.traits` and manually remove `trait_added` and `trait_modified` from the dictionary instead of using the `HasTraits.editable_traits` method. For example, this: https://github.com/enthought/mayavi/blob/e896f99af84c81052d048d4ae03f245b9d27ee83/tvtk/tools/tvtk_doc.py#L119-L120

`*_items` added for supporting `on_trait_change` with `List`, `Set`, `Dict` are internal traits, just like `trait_added` and `trait_modified`. But traits that are added by the user and happen to have `"_items"` as a suffix are not internal traits.

In short, `HasTraits.traits` report some internal traits but not some others. However this method is likely widely used that it would be difficult to change without breaking downstream code.

Perhaps the solution to this issue is to grow a new method?

**Comment by kitchoi** [comment]
If I proceed with the proposal I made above and run Envisage test suite with that, I get these two failures:
```
======================================================================
FAIL: test_append (envisage.tests.test_extension_point_changed.ExtensionPointChangedTestCase)
append
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/kchoi/Work/ETS/envisage/envisage/tests/test_extension_point_changed.py", line 90, in test_append
    self.assertEqual(7, len(extensions))
AssertionError: 7 != 6

======================================================================
FAIL: test_remove (envisage.tests.test_extension_point_changed.ExtensionPointChangedTestCase)
remove
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/kchoi/Work/ETS/envisage/envisage/tests/test_extension_point_changed.py", line 131, in test_remove
    self.assertEqual(5, len(extensions))
AssertionError: 5 != 6
```
While I know what I changed that caused the failure, I had not yet worked out _how_ the failure happens. In any case, I am afraid the method's behaviour maybe sealed :/

**Comment by kitchoi** [comment]
@mdickinson What do you think about the behaviour of `HasTraits.traits` being sealed and that we should close this as wont-fix?

New context today: I am trying to fix https://github.com/enthought/apptools/issues/129 and I realize I could not filter "_items" traits that are created for List/Set/Dict mutation events in general. So I am thinking about opening an issue to enable that filtering. There is some overlap with this issue here.

**Comment by mdickinson** [comment]
> What do you think about the behaviour of `HasTraits.traits` being sealed and that we should close this as wont-fix?

Yep, sadly I think this is something that's baked-in enough that it's too high-risk to change - I've seen a lot of code over the years that parses the results from `HasTraits.traits` for one reason or another.

**Comment by kitchoi** [comment]
Thank you. Agreed. Closing.

**Comment by kitchoi** [comment]
> New context today: I am trying to fix enthought/apptools#129 and I realize I could not filter "_items" traits that are created for List/Set/Dict mutation events in general. So I am thinking about opening an issue to enable that filtering. There is some overlap with this issue here.

FTR, I was being stupid - I found a different way to solve this.
The desirable behaviour of skipping "_items" traits added by Traits may still apply, and I commented on an existing issue #1098 instead.
